File: back-end/src/main/java/es/ull/project/application/usecase/infrastructureplan/CreateInfrastructurePlanUseCase.java
package es.ull.project.application.usecase.infrastructureplan;

import java.util.List;
import java.util.UUID;

import es.ull.project.domain.entity.InfrastructurePlan;
import es.ull.project.domain.valueobject.cost.MaximumBudget;
import es.ull.project.domain.valueobject.policy.ServicePolicies;
import es.ull.project.domain.valueobject.time.PlanningPeriod;

/**
 * Use case for creating a new infrastructure plan.
 */
public interface CreateInfrastructurePlanUseCase {
    /**
     * Creates a new infrastructure plan with the specified attributes.
     *
     * @param period          the planning period
     * @param maxBudget       the maximum budget
     * @param servicePolicies the service policies
     * @param selectedFacilityIds list of facility IDs to include in the plan
     * @param serviceAssignmentIds list of service assignment IDs to include in the plan
     * @return the created infrastructure plan
     */
    InfrastructurePlan create(
        PlanningPeriod period, 
        MaximumBudget maxBudget, 
        ServicePolicies servicePolicies,
        List<UUID> selectedFacilityIds,
        List<UUID> serviceAssignmentIds
    );
}

File: front-end/src/application/usecase/InfrastructurePlanManagement/create-infrastructure-plan-use-case.ts
// CreateInfrastructurePlanUseCase.ts
// Use case contract for creating a new infrastructure plan

import type { Either, DataError } from '@ull-tfg/ull-tfg-typescript';
import { InfrastructurePlan } from '../../../domain/entity/infrastructure-plan';
import { PlanningPeriod } from '../../../domain/valueobject/time/planning-period';
import { MaximumBudget } from '../../../domain/valueobject/cost/maximum-budget';
import { ServicePolicies } from '../../../domain/valueobject/policy/service-policies';

/**
 * Use case for creating a new infrastructure plan in the system.
 * Input: period, maxBudget, servicePolicies (optional)
 * Output: InfrastructurePlan (entity)
 */

// Command object for input data
export interface CreateInfrastructurePlanCommand {
    period: PlanningPeriod;
    maxBudget: MaximumBudget;
    servicePolicies?: ServicePolicies | null;
}

// Result type for the use case
export type CreateInfrastructurePlanResult = InfrastructurePlan;

// Use case contract
export interface CreateInfrastructurePlanUseCase {
    /**
     * Handles the creation of a new infrastructure plan
     *
     * @param command Data needed to create the infrastructure plan
     * @returns Either a DataError or the created InfrastructurePlan
     */
    execute(command: CreateInfrastructurePlanCommand): Promise<Either<DataError, CreateInfrastructurePlanResult>>;
}
